### ISSUE TITLE
support for packages installed with --use-pep517

### ISSUE THREAD

**Issue by karlicoss** [seanbreckenridge/reorder_editable#2]
I encountered an issue here https://github.com/pypa/setuptools/issues/4059, and one of the suggestions was to use `--use-pep517` flag. I checked it out, and it seems it's not using `easy-install.pth` anymore, instead creating a bunch of files like these

```
__editable__.HPI-0.3.20230328.dev34+dirty.pth
__editable__.HPI_private-0.0.0.pth
__editable__.reorder_editable-0.1.0.pth
__editable___HPI_0_3_20230328_dev34_dirty_finder.py
__editable___reorder_editable_0_1_0_finder.py
```

Haven't found yet what controls namespace packages import order under that model (or if this pep517 is even anything to worry about?), so figured I'd just leave it here for now

**Comment by seanbreckenridge** [comment]
Thanks for letting me know, will watch the other issue to see if there could be changes made here

**Comment by seanbreckenridge** [comment]
Interesting that the PEP doesnt seem to answer the question?

<https://peps.python.org/pep-0517/>

Wonder if there are docs on how this is working:

> Note
> 
> Editable installs
> 
> This PEP originally specified another hook, install_editable, to do an editable install (as with pip install -e). It was removed due to the complexity of the topic, but may be specified in a later PEP.
> 
> Briefly, the questions to be answered include: what reasonable ways existing of implementing an ‘editable install’? Should the backend or the frontend pick how to make an editable install? And if the frontend does, what does it need from the backend to do so.
>

**Comment by abravalheri** [comment]
`pip`'s flag naming is misleading. `--use-pep517` actually means "use build isolation", with "build isolation" being defined as the process to create `sdist`/`wheel` artifacts using a brand new virtual environment + a series of backend APIs[^1].

The series of backend APIs are defined by PEP 517, PEP 660. But that may include any other PEPs in the future.

`pip` is supposed to use "build isolation" by default at some point in the future (it was meant to be in 23.1, but it got postponed).

`pip` will automatically opt into "build isolation" in some circumstances:

- The `pyproject.toml` file is present in the project root (I am not sure, but I think it can even be an empty file)
- The user passes `--use-pep517`
- The user does not have `setuptools` and `wheel` installed in their environment.

---

Before the standards on build isolation, `setuptools` would place an `easy-install.pth` file and a series of other `*.egg-link` files on `site-packages`.

With the new standard out, this is no longer an option. Effectively setuptools cannot interact with the `site-packages` directly and instead needs to create an 'editable wheel'.
This editable wheel contains a new set of files (those `__editable__*` files), which are meant to be a better approximation for the real installation[^2], from the point of view of packaging[^3].

[^1]: You can still use the backend APIs without virtual environment isolation by combining `--use-pep517 --no-build-isolation`.
[^2]: Any editable installation method will still have limitations and never fully mimic the real one.
[^3]: Static analysis tools may have problems to cope with dynamic import behaviour (which is a perfectly valid feature of the Python language, but understandably are hard to digest during static analysis).

**Comment by seanbreckenridge** [comment]
Did a test in a venv with all 3 of my `HPI` namespace packages installed with build isolation, I see these when starting in `-v` mode:

```
import '__editable___HPI_0_3_20230328_dev37_finder' # <_frozen_importlib_external.SourceFileLoader object at 0x7f6576c061d0>
Processing .pth file: '/tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__editable__.HPI_seanbreckenridge-0.0.1.pth'
# /tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__pycache__/__editable___HPI_seanbreckenridge_0_0_1_finder.cpython-311.pyc matches /tmp/m8ZFkz4Oqn/venv/lib/python3.11/sit
e-packages/__editable___HPI_seanbreckenridge_0_0_1_finder.py
# code object from '/tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__pycache__/__editable___HPI_seanbreckenridge_0_0_1_finder.cpython-311.pyc'
import '__editable___HPI_seanbreckenridge_0_0_1_finder' # <_frozen_importlib_external.SourceFileLoader object at 0x7f6576b54710>
Processing .pth file: '/tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__editable__.HPI_seanbreckenridge_personal-0.1.0.pth'
# /tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__pycache__/__editable___HPI_seanbreckenridge_personal_0_1_0_finder.cpython-311.pyc matches /tmp/m8ZFkz4Oqn/venv/lib/pytho
n3.11/site-packages/__editable___HPI_seanbreckenridge_personal_0_1_0_finder.py
# code object from '/tmp/m8ZFkz4Oqn/venv/lib/python3.11/site-packages/__pycache__/__editable___HPI_seanbreckenridge_personal_0_1_0_finder.cpython-311.pyc'
import '__editable___HPI_seanbreckenridge_personal_0_1_0_finder' # <_frozen_importlib_external.SourceFileLoader object at 0x7f6576b56010>
```

Traced it to the `site` module and this seems to be what is determining the order:

<https://github.com/python/cpython/blob/65c285062ce2769249610348636d3d73153e0144/Lib/site.py#L207-L229>

Its a:

```
    for name in sorted(names):
        addpackage(sitedir, name, known_paths)
```

That hasnt been changed for 15 years, so its probably something we could depend on

**Comment by karlicoss** [comment]
Seems like the only option is to rename the packages, e.g. `hpi-main -> 10_hpi-main`, `hpi-overlay -> 00_hpi-overlay`, so they end up in the expected order when sorted.

But it's kinda annoying since that requires modifying the source code (so your local copy will end up always dirty).

Spent some time in `setuptools`/`pip` internals trying to understand if there is an easy way of overriding that dynamically (e.g. by specifying a pip flag or something, but I'm kinda lost :( )

So far the easiest way I think is:

- install with `--use-pep517 --no-build-isolation`. The latter prevents creating extra copy of `.egg_info` inside the source directory (typically in `src/` dir) -- mainly did it because I'm pretty confused how it interacts with `.dist-info` inside user's site
- rename the `__editable__*` files in the user site:

  ```
  ~/.local/lib/python3.10/site-packages $ ls | grep pth | grep hpi
  __editable__.00_hpi_overlay-0.0.0.pth
  __editable__.10_hpi_main-0.0.0.pth
  ```
  Just this seems to be already enough for it to work in runtime
  
   ```
   $ python3 -c 'import my; print(my.__path__)'
  _NamespacePath(['/project/overlay/src/my', '/project/main/src/my'])
  ```

   The only downside is that if we `pip uninstall` the package, it would leave the renamed `__editable__` file behind. To work around that we could do a couple of extra steps:

- rename the dist-info dirs `~/.local/lib/python3.10/site-packages $ mv {,10_}hpi_main-0.0.0.dist-info` && mv {,00_}hpi_overlay-0.0.0.dist-info`
- rename the package in the metadata file

  ```
   sed -i 's/hpi-overlay/00-hpi-overlay/' 00_hpi_overlay-0.0.0.dist-info/METADATA 
   sed -i 's/hpi-main/10-hpi-main/' 10_hpi_main-0.0.0.dist-info/METADATA
   ```

  After that, the renamed package properly shows up in pip:

  ```
  $ pip3 list | grep hpi
  00-hpi-overlay     0.0.0       /project/overlay
  10-hpi-main        0.0.0       /project/main
  ```

  And uninstalling should cleanup all relevant files. 

Pretty elaborate, but I guess all of this can be scripted :shrug: 

The downside is that you need to make sure you editable install the package only once, otherwise you need to rerun =reorder_editable= (but that's no different from the old way). 
The only difference now is that if you install the package again, it's gonna end up with the original name (i.e. `hpi-main/hpi-overlay`), and now you'll end up with extra copies. The tool could probably sort that out automatically too, but going to be trickier than simply editing a text file :(

**Comment by karlicoss** [comment]
Ok, after doing all that (and remembering some stuff I read about loading `.pth` files), I realized maybe there is a much simpler way:

Create a `.pth` file so that it appears before all other `__editable__` files:

```
~/.local/lib/python3.10/site-packages $ ls -al | grep pth
-rw-r--r--  1 adhoc adhoc       39 Dec 22 23:15 __00_force_hpi_order.pth
-rw-r--r--  1 adhoc adhoc      151 Dec 22 21:35 distutils-precedence.pth
-rw-r--r--  1 adhoc adhoc        0 Dec 22 23:16 easy-install.pth
-rw-r--r--  1 adhoc adhoc       18 Dec 22 23:17 __editable__.hpi_main-0.0.0.pth
-rw-r--r--  1 adhoc adhoc       21 Dec 22 23:17 __editable__.hpi_overlay-0.0.0.pth
```

Put the desired editable order in there:
```
$ cat __00_force_hpi_order.pth 
/project/overlay/src
/project/main/src
```

After that it all just works as expected without renaming packages etc; and also works regardless  whether `--use-pep517` is used, which is nice since means the code for `reorder_editable` can stay simple, just need to change the file it manages from `easy-install.pth`

The only downside I see is that imports from packages mentioned in this file are going to work even if the package isn't properly installed with `pip` which may be kinda confusing.

**Comment by seanbreckenridge** [comment]
Ah yeah, that would likely work. Will think about how to create some new command which lets you specify a particular custom `.pth` file to use/edit.

Thanks for doing the dive into the source code :+1: , I could barely understand anything except it loading the .pth files in order

**Comment by karlicoss** [comment]
Aw man now it got even more confusing... 

So I've been setting up my hpi environment after reinstalling the system with all latest python/pip/setuptools.
Reproducing in docker just for the demo:

```
docker run --rm -it python:3.12 /bin/bash

$ pip install -U setuptools pip
$ cd /tmp/
$ git clone https://github.com/karlicoss/HPI.git
$ git clone https://github.com/seanbreckenridge/HPI.git hpi_sean
$ pip install --user --editable /tmp/HPI 
```

So this creates even weirder `__editable__` files than before:

```
$ ls -al /root/.local/lib/python3.12/site-packages/
drwxr-xr-x 2 root root 4096 Sep  9 21:46 HPI-0.5.20240825.dev27.dist-info
-rw-r--r-- 1 root root  103 Sep  9 21:46 __editable__.HPI-0.5.20240825.dev27.pth
-rw-r--r-- 1 root root 6127 Sep  9 21:46 __editable___HPI_0_5_20240825_dev27_finder.py
```

The `.pth` files imports this finder thingie:
```
$ cat /root/.local/lib/python3.12/site-packages/__editable__.HPI-0.5.20240825.dev27.pth
import __editable___HPI_0_5_20240825_dev27_finder; __editable___HPI_0_5_20240825_dev27_finder.install()
```

The *finder.py file is generated based on this themplate :exploding_head: https://github.com/pypa/setuptools/blob/962835d068af4d993f2195e7a6ffd4f70548b9f4/setuptools/command/editable_wheel.py#L801
In particular it contains this huge mapping with each subpackage listed along with the path:

```
...
MAPPING: dict[str, str] = {'my': '/tmp/HPI/my'}
NAMESPACES: dict[str, list[str]] = {'my': ['/tmp/HPI/my'], 'my.time': ['/tmp/HPI/my/time'], 'my.bumble': ['/tmp/HPI/my/bumble'], 'my.calendar': ['/tmp/HPI/my/calendar'],...
...
```


Even without any extra hacks like `reorder_editable` it already results multiple paths in `__path__` :thinking: 
```
$ python3 -c 'import my; print(my.__path__)'
_NamespacePath(['/tmp/HPI/my', '__editable__.HPI-0.5.20240825.dev27.finder.__path_hook__'])
```


Now if we install another overlay:

```
$ pip install --user --editable /tmp/hpi_sean
$ python3 -c 'import my; print(my.__path__)'
_NamespacePath(['/tmp/HPI/my', '__editable__.HPI-0.5.20240825.dev27.finder.__path_hook__', '/tmp/hpi_sean/my', '__editable__.HPI_seanbreckenridge-0.0.1.finder.__path_hook__'])
```

So the paths are in the wrong order... creating a `__00_force_hpi_order.pth` as I suggested above seems to solve it though.

```
$ cat /root/.local/lib/python3.12/site-packages/__00_force_hpi_order.pth 
/tmp/hpi_sean
/tmp/HPI
```

... with a ton of extra crap in `__path__` though
```
$ python3 -c 'import my; print(my.__path__)'
_NamespacePath(['/tmp/hpi_sean/my', '/tmp/HPI/my', '/tmp/HPI/my', '__editable__.HPI-0.5.20240825.dev27.finder.__path_hook__', '/tmp/hpi_sean/my', '__editable__.HPI_seanbreckenridge-0.0.1.finder.__path_hook__'])
```
